### PR TITLE
langchain: Fix broken `OpenAIModerationChain` and implement async

### PR DESCRIPTION
**Pull request by mattflo**
Thank you for contributing to LangChain!

## PR title
lancghain[patch]: fix `OpenAIModerationChain` and implement async

## PR message
Description: fix `OpenAIModerationChain` and implement async

Issues: 
- https://github.com/langchain-ai/langchain/issues/18533 
- https://github.com/langchain-ai/langchain/issues/13685

Dependencies: none
Twitter handle: mattflo

## Add tests and docs
 
Existing documentation is broken: https://python.langchain.com/docs/guides/safety/moderation

- [ x] **Lint and test**: Run `make format`, `make lint` and `make test` from the root of the package(s) you've modified. See contribution guidelines for more: https://python.langchain.com/docs/contributing/

### CODE REVIEW

File: libs/langchain/langchain/chains/moderation.py
Comment (baskaryan) [inline libs/langchain/langchain/chains/moderation.py:1]: if possible it'd be nice to make this backwards compatible with openai v0, the way openai models are

File: libs/langchain/langchain/chains/moderation.py [hunk 1]
```diff
@@ -76,8 +88,14 @@ def output_keys(self) -> List[str]:
         """
         return [self.output_key]
 
-    def _moderate(self, text: str, results: dict) -> str:
-        if results["flagged"]:
+    def _moderate(self, text: str, results: Any) -> str:
+        if self.openai_version is None:
+            raise AssertionError("Got None version type.")
```
Comment (ccurme) [inline libs/langchain/langchain/chains/moderation.py:93]: these checks are just for mypy, open to suggestions on better ways to do this.

File: libs/langchain/langchain/chains/moderation.py [hunk 2]
```diff
@@ -76,8 +88,14 @@ def output_keys(self) -> List[str]:
...
+        if self.openai_version < "1.0.0":
```
Comment (efriis) [inline libs/langchain/langchain/chains/moderation.py:94]: does this string comparison work or do we need to convert to a version object?

File: libs/langchain/langchain/chains/moderation.py [hunk 3]
```diff
@@ -33,6 +38,7 @@ class OpenAIModerationChain(Chain):
     output_key: str = "output"  #: :meta private:
     openai_api_key: Optional[str] = None
     openai_organization: Optional[str] = None
+    openai_version: Optional[str] = None
```
Comment (efriis) [inline libs/langchain/langchain/chains/moderation.py:41]: if you want you can also do this, and then you don't need the mypy stuff below (validation guarantees it's populated)
```suggestion
    openai_version: str = Field(default=None)
```

**Review by efriis — APPROVED** [review]
1 note otherwise looks good

**Pull request merged**